worker/provisioner/provisioner_task.go

```diff
@@ -689,106 +714,357 @@ func (task *provisionerTask) maintainMachines(machines []*apiprovisioner.Machine
+		dgMachines[i] = machine.MachineTag()
+	}
+	distributionGroups, err := task.distributionGroupFinder.DistributionGroupByMachineId(dgMachines...)
+	if err != nil {
+		return machineZones, err
+	}
+
+	task.azMachinesMutex.Lock()
+	defer task.azMachinesMutex.Unlock()
+
+	// assign an initial az to a machine based on lowest population.
+	// if the machine has a distribution group, assign based on lowest
+	// az population of the distribution group machines.
+	for i, machine := range machines {
+		dgResult := distributionGroups[i]
+		if dgResult.Err != nil {
+			return machineZones, err
 		}
-		possibleTools, err := task.toolsFinder.FindTools(
-			*v,
-			pInfo.Series,
-			arch,
-		)
-		if err != nil {
-			return task.setErrorStatus("cannot find tools for machine %q: %v", m, err)
+		if len(dgResult.MachineIds) > 0 {
+			dgZoneMap := task.populateDistributionGroupZoneMap(dgResult.MachineIds)
+			sort.Sort(byPopulationThenName(dgZoneMap))
+
+			for _, dgZoneMachines := range dgZoneMap {
+				if dgZoneMachines.failedMachineIds.Contains(machine.Id()) == false {
+					machineZones[machine] = dgZoneMachines.zoneName
+					for _, azm := range task.availabilityZoneMachines {
+						if azm.zoneName == dgZoneMachines.zoneName {
+							azm.machineIds.Add(machine.Id())
+							break
+						}
+					}
+					break
+				}
+			}
+		} else {
+			sort.Sort(byPopulationThenName(task.availabilityZoneMachines))
+			for _, zoneMachines := range task.availabilityZoneMachines {
+				if zoneMachines.failedMachineIds.Contains(machine.Id()) == false {
+					machineZones[machine] = zoneMachines.zoneName
+					zoneMachines.machineIds.Add(machine.Id())
+					break
+				}
+			}
 		}
+	}
+	return machineZones, nil
+}
 
-		startInstanceParams, err := constructStartInstanceParams(
-			task.controllerUUID,
-			m,
-			instanceCfg,
-			pInfo,
-			possibleTools,
-		)
-		if err != nil {
-			return task.setErrorStatus("cannot construct params for machine %q: %v", m, err)
+type byPopulationThenName []*availabilityZoneMachine
+
+func (b byPopulationThenName) Len() int {
+	return len(b)
+}
+
+func (b byPopulationThenName) Less(i, j int) bool {
+	switch {
+	case b[i].machineIds.Size() < b[j].machineIds.Size():
+		return true
+	case b[i].machineIds.Size() == b[j].machineIds.Size():
+		return b[i].zoneName < b[j].zoneName
+	}
+	return false
+}
+
+func (b byPopulationThenName) Swap(i, j int) {
+	b[i], b[j] = b[j], b[i]
+}
+
+// startMachines starts a goroutine for each specified machine to
+// start it.  The specified machines are provided the name of the
+// first availability zone to try.  Errors from individual start
+// machine attempts will be logged.
+func (task *provisionerTask) startMachines(machines []*apiprovisioner.Machine) error {
+	if len(machines) == 0 {
+		return nil
+	}
+
+	// Get first attempt mapping of machine to availability zone, not all clouds
+	// have availability zones, so error here is not fatal.
+	machineZones, _ := task.machineAvailabilityZoneDistribution(machines)
```

Comment:
I think we can make this call to machineAvailabilityZoneDistribution go away, and just make calls to it in startMachine. Then machineAvailabilityZoneDistribution should only need to take a single machine, and return a single zone.

In startMachine, you would do something like:

```go
zone, err := machineAvailabilityZoneDistribution(machine)
...
attemptsLeft := ...
for attemptsLeft > 0 {
    ..., err := env.StartInstance(args)
    if err == environs.ErrAvailabilityZoneFailed {
        zone, err := machineAvailabilityZoneDistribution(machine)
        ...
        attemptsLeft = ... // reset attemptsLeft, for the new AZ
    }
}
```

finally, machineAvailabilityZoneDistribution could then be made to return an error if there are no remaining AZs to attempt for the machine

Reply:
This will make adding PreviewAvailabilitySuccess() into the code easier too.  Thank you!

Reply:
FWIW, I came up with a different name for the ZonedEnviron method: DeriveAvailabilityZone.

```go
type ZonedEnviron interface {
    ...
    // DeriveAvailabilityZone attempts to derive an availability zone from
    // the specified StartInstanceParams.
    //
    // The parameters for starting an instance may imply (or explicitly
    // specify) an availability zone, e.g. due to placement, or due to the
    // attachment of existing volumes. If there is no such restriction,
    // then DeriveAvailabilityZone should return an empty string to
    // indicate that the caller should choose an availability zone.
    DeriveAvailabilityZone(StartInstanceParams) (string, error)
    ...
}
```